Check `fut_gid` against `run_gid` in addition to supplemental groups

philipsd6 opened the pull request:
##### SUMMARY
Unarchiving a ZipArchive into an existing destination triggers a check to ensure that the user and group ownership can be set. The target user is checked against the running user, but the target group is only checked against the supplemental groups, which fails if the target group is the running primary group. This PR adds the checks against the `run_group` and `run_git` .

Fixes #49284

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
unarchive module

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
See https://github.com/ansible/ansible/issues/49284#issuecomment-563334848 for a description of a playbook that fails in latest AWX using Ansible 2.10.0.dev0.
<!--- Paste verbatim command output below, e.g. before and after your change -->

philipsd6 committed:
Check `fut_gid` against `run_gid` in addition to supplemental groups

Signed-off-by: Philip Douglass <philip.douglass@amadeus.com>

philipsd6 commented:
The Shippable — Run 153181: "Failed to validate the SSL certificate for releases.ansible.com" appears to be irrelevant to this PR.

samdoran reviewed (changes requested):
This fix seems correct.

Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) and integration tests. See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

philipsd6 commented:
> This fix seems correct.
> 
> Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) and integration tests. See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

@samdoran, I added the changelog fragment as requested, but it's not clear how to write an integration test that will properly test this. The integration tests already include a task that unarchives the same zipfile twice, but this issue is only triggered if the target user is non-root, and the tests are run as root in the docker container.

samdoran commented:
@philipsd6 You could add tasks to create a standard user, then user `become` and `become_user` to run tasks as that user. There are [some examples](https://github.com/ansible/ansible/blob/8ae5e34127216bf56544ccfd54cc48979741ad48/test/integration/targets/become/tasks/sudo.yml) in other integration tests.

philipsd6 committed:
Add integration tests for unarchiving as unprivileged user

- Skip on MacOSX as the test is not validated on that platform

Signed-off-by: Philip Douglass <philip.douglass@amadeus.com>

philipsd6 commented:
@samdoran Good idea, but I cannot seem to reproduce the original error with an integration test, because when I create a normal user with the `user` module, `os.getgroups()` has the primary group in the list, which then works in the `unarchive` module as is. However, in the AWX container, the `os.getgroups()` call returns an empty list, which is what causes the `unarchive` module to fail, which this PR fixes.

In AWX container:
```bash
philipsd6 /d/awx> docker exec -it tools_awx_1 bash
bash-4.4$ /venv/awx/bin/python -c "import os;print(os.getuid(), os.getgroups())"
74892 []
```

With `ansible-test integration --docker centos8 -v unarchive` and the following task:

```yaml
  - command: '{{ansible_playbook_python}} -c "import os;print(os.getuid(), os.getgroups())"'
```

results:
```
TASK [unarchive : command] *****************************************************
changed: [testhost] => {"changed": true, "cmd": ["/usr/bin/python3.6", "-c", "import os;print(os.getuid(), os.getgroups())"], "delta": "0:00:00.030087", "end": "2020-01-18 18:32:55.038357", "rc": 0, "start": "2020-01-18 18:32:55.008270", "stderr": "", "stderr_lines": [], "stdout": "1000 [1000]", "stdout_lines": ["1000 [1000]"]}
```

Note `"stdout": "1000 [1000]"`. If I could figure out how to create a normal user with an empty supplementary groups list as reported by `os.getgroups()` I would be able to write an integration test.

philipsd6 commented:
@samdoran Good idea, but I cannot seem to reproduce the original error with an integration test, because when I create a normal user with the `user` module, `os.getgroups()` has the primary group in the list, which then works in the `unarchive` module as is. However, in the AWX container, the `os.getgroups()` call returns an empty list, which is what causes the `unarchive` module to fail, which this PR fixes.

In AWX container:
```bash
philipsd6 /d/awx> docker exec -it tools_awx_1 bash
bash-4.4$ /venv/awx/bin/python -c "import os;print(os.getuid(), os.getgroups())"
74892 []
```

With `ansible-test integration --docker centos8 -v unarchive` and the following task:

```yaml
  - command: '{{ansible_playbook_python}} -c "import os;print(os.getuid(), os.getgroups())"'
```

results:
```
TASK [unarchive : command] *****************************************************
changed: [testhost] => {"changed": true, "cmd": ["/usr/bin/python3.6", "-c", "import os;print(os.getuid(), os.getgroups())"], "delta": "0:00:00.030087", "end": "2020-01-18 18:32:55.038357", "rc": 0, "start": "2020-01-18 18:32:55.008270", "stderr": "", "stderr_lines": [], "stdout": "1000 [1000]", "stdout_lines": ["1000 [1000]"]}
```

Note `"stdout": "1000 [1000]"`. If I could figure out how to create a normal user with an empty supplementary groups list as reported by `os.getgroups()` I would be able to write an integration test.

samdoran commented on test/integration/targets/unarchive/tasks/main.yml:
```diff
@@ -659,3 +659,50 @@
+- name: Create test user with root primary group
+  user:
+    name: unarchivetest1
+    group: root
```
From looking at the test failure, the `root` group doesn't exist on FreeBSD or macOS. Maybe use a different group for those systems. Look at `default_user_group` in `vars/main.yml` as an example.

Alternatively, we could skip this test on FreeBSD and macOS by adding a conditional to this block since we have enough coverage on other platforms.

philipsd6 replied:
I'll choose to skip this test on FreeBSD and macOS then, as there's nothing in the fix that references root, this is just to demonstrate the fix works, and as you say we have enough coverage.

philipsd6 replied:
It's not necessary to create the test user as a root user, so I removed that option entirely.

samdoran commented on test/integration/targets/unarchive/tasks/main.yml:
```diff
@@ -659,3 +659,50 @@
+- name: Test unarchiving twice as test user
+  block:
+  - name: create our zip unarchive destination as test user
+    file: path={{remote_tmp_dir}}/test-unarchive-zip state=directory
```
Please use multi-line YAML for new tests.
```suggestion
    file: 
      path: "{{ remote_tmp_dir }}/test-unarchive-zip"
      state: directory
```

philipsd6 replied:
I will update my branch and rebase it with this suggestion.

samdoran commented:
> If I could figure out how to create a normal user with an empty supplementary groups list as reported by `os.getgroups()` I would be able to write an integration test.

Could you use a `command` task to either create user without supplemental groups, or remove supplemental groups after the `user` task? It may be easier to only run on Linux in that case so you don't have to work out a ton of commands. Or you could use a dictionary to lookup the correct command per OS.

philipsd6 commented:
OK, here's the status of this PR:

1. it is impossible to create a test to replicate the environment in the AWX dev image that causes this module to fail. Even creating the test user by writing a line to `/etc/passwd` like the awx entrypoint script does doesn't end up causing `os.getgroups()` to return an empty list. (I suspect this is particular to the way the AWX docker container is instantiated.) So I just made the test run through the paces as an unprivileged user for this particular use-case.

2. The integration test I previously provided was a mess (`become: unarchivetest1`? What was I thinking?) The current one works as expected, but it doesn't fail without my fix, because `os.getgroups()` doesn't return an empty list. So it basically proves nothing, except:

3. `unarchive` can run twice correctly on any Posix system as an unprivileged user _except_ OSX. On OSX, the Shippable logs show that the second `unarchive` results in `changed: yes` when it should report `changed: no` as on all the other systems. This is a bug exposed by this new integration test, even though the PR was not intended to address this.

So what shall we do? I _could_ skip OSX, but the_ failed check on OSX does appear to be a valid issue, even if irrelevant to this PR.

I _have_ confirmed that with this fix, I can properly unarchive a zipfile into a symlink in the AWX container, whereas without it I cannot, as it raises the `Cannot change group ownership of...` exception.

philipsd6 commented:
ready_for_review

1 down, 5 to go...

jpdias92 commented:
I would just like to thank @philipsd6  for this fix and bring to attention that more people are needing this fix to get merged.

It would be great if this fix was shipped with an Ansible release in the near future...

We are already using this fix in my team, by changing the file when building our Docker image where we run Ansible, but it would be much cleaner if this was already part of an Ansible release.

Thanks!

samdoran commented:
This is failing tests on macOS. That needs to be resolved before this can be merged.

philipsd6 commented:
> This is failing tests on macOS. That needs to be resolved before this can be merged.

@samdoran, as I said back on Feb 29, I know the integration test is failing on macOS X, but this is because the test exposes an existing issue this PR was not intended to fix. Since I do not have a macOS X system to test on, I cannot fix this, so I will just adjust the test to ignore that platform -- but to be clear, this will just be a superficial fix.

philipsd6 commented:
/rebuild_failed

philipsd6 commented:
Shippable looks good now -- hope we can get this merged finally @samdoran!

samdoran committed:
Bring some sanity to the unarchive tests

Break tasks into separat files for easier reading and maintenance

samdoran committed:
Fix tests for macOS

Creating a user by specifying a default group of 'staff'
The user module does not actually remove the user directory on macOS,
so explicitly remove it.
Put the removal tasks in an always block to ensure they always run

samdoran commented:
I figured out how to get the tests passing on macOS. I will push up my fix shortly.

samdoran committed:
Remove comment

samdoran committed:
Remove debugging cruft
